stdlib/bigstring.ml

```diff
@@ -0,0 +1,199 @@
+  else Bytes.unsafe_to_string (unsafe_sub_bytes b pos len)
+
+let of_bytes b =
+  let len = Bytes.length b in
+  let b' = create len in
+  unsafe_blit_from_bytes b 0 b' 0 len;
+  b'
+
+let to_bytes b =
+  let len = length b in
+  let b' = Bytes.create len in
+  unsafe_blit_to_bytes b 0 b' 0 len;
+  b'
+
+let of_string s =
+  of_bytes (Bytes.unsafe_of_string s)
+
+let to_string b =
+  Bytes.unsafe_to_string (to_bytes b)
+
+let to_seq b =
+  let rec loop sz i () =
+    if i = sz then
+      Seq.Nil
+    else
+      let x = unsafe_get b i in
+      Seq.Cons (x, loop sz (succ i))
+  in
+  loop (length b) 0
+
+external get_uint16_ne: t -> int -> int = "%caml_bigstring_get16"
+external get_int32_ne: t -> int -> int32 = "%caml_bigstring_get32"
+external get_int64_ne: t -> int -> int64 = "%caml_bigstring_get64"
+
+external set_int16_ne: t -> int -> int -> unit = "%caml_bigstring_set16"
+external set_int32_ne: t -> int -> int32 -> unit = "%caml_bigstring_set32"
+external set_int64_ne: t -> int -> int64 -> unit = "%caml_bigstring_set64"
+
+external swap16: int -> int = "%bswap16"
+external swap32: int32 -> int32 = "%bswap_int32"
+external swap64: int64 -> int64 = "%bswap_int64"
+
+let get_uint8 b i =
+  Char.code (get b i)
+
+let get_int8 b i =
+  ((get_uint8 b i) lsl (Sys.int_size - 8)) asr (Sys.int_size - 8)
+
+let get_uint16_le b i =
+  if Sys.big_endian then swap16 (get_uint16_ne b i)
+  else get_uint16_ne b i
+
+let get_uint16_be b i =
+  if not Sys.big_endian then swap16 (get_uint16_ne b i)
+  else get_uint16_ne b i
+
+let get_int16_ne b i =
+  ((get_uint16_ne b i) lsl (Sys.int_size - 16)) asr (Sys.int_size - 16)
+
+let get_int16_le b i =
+  ((get_uint16_le b i) lsl (Sys.int_size - 16)) asr (Sys.int_size - 16)
+
+let get_int16_be b i =
+  ((get_uint16_be b i) lsl (Sys.int_size - 16)) asr (Sys.int_size - 16)
+
+let get_int32_le b i =
+  if Sys.big_endian then swap32 (get_int32_ne b i)
+  else get_int32_ne b i
+
+let get_int32_be b i =
+  if not Sys.big_endian then swap32 (get_int32_ne b i)
+  else get_int32_ne b i
+
+let get_int64_le b i =
+  if Sys.big_endian then swap64 (get_int64_ne b i)
+  else get_int64_ne b i
+
+let get_int64_be b i =
+  if not Sys.big_endian then swap64 (get_int64_ne b i)
+  else get_int64_ne b i
+
+let set_int8 b i x =
+  set b i (Char.unsafe_chr x)
+
+let set_int16_le b i x =
+  if Sys.big_endian then set_int16_ne b i (swap16 x)
+  else set_int16_ne b i x
+
+let set_int16_be b i x =
+  if not Sys.big_endian then set_int16_ne b i (swap16 x)
+  else set_int16_ne b i x
+
+let set_int32_le b i x =
+  if Sys.big_endian then set_int32_ne b i (swap32 x)
+  else set_int32_ne b i x
+
+let set_int32_be b i x =
+  if not Sys.big_endian then set_int32_ne b i (swap32 x)
+  else set_int32_ne b i x
+
+let set_int64_le b i x =
+  if Sys.big_endian then set_int64_ne b i (swap64 x)
+  else set_int64_ne b i x
+
+let set_int64_be b i x =
+  if not Sys.big_endian then set_int64_ne b i (swap64 x)
+  else set_int64_ne b i x
```

Comment:
Could these be written as:

```ocaml
let set_int64_be =
   if not Sys.big_endian then set_int64_swap
   else set_int64_ne
```

I.e., moving the endianness check to the top-level?

Reply:
Yes, they could, but I don't think it amounts to anything: under `ocamlopt` both variants generate the same code thanks to inlining. And even under `ocamlc` you are trading a branch on `Sys.big_endian` for an extra function call to `set_int64_swap`.

Reply:
In this particular case, the two ways to write the code give exactly the same result: the test is reduced at compile-time, and the branches in your version are external primitives, so they are eta-expanded when they are not applied.